src/api/base.ts

```diff
@@ -68,58 +32,127 @@ const apiAxiosInstance = axios.create({
+  const options: RequestOptions = {
+    skipLoader: false,
+    withDelay: 0,
+    ...opts,
+  }
+
+  const wait = window.setTimeout(() => {
+    if (options.skipLoader !== true) {
+      useAppSpinner.show()
+    }
+  }, 200)
+
+  return apiAxiosInstance({ ...config, ...opts })
+    .then((success: AxiosResponse<T>) => {
+      return success.data
+    })
+    .catch((err: unknown) => {
+      if (axios.isAxiosError(err)) {
+        throw new HttpError(
+          err.message,
+          err.response?.status,
+          err.response?.data,
+          axios.isCancel(err) ? HTTP_CANCELLED_ERROR : HTTP_ERROR
+        )
       }
-    }, 200)
-
-    return apiAxiosInstance({ ...config, ...opts })
-      .then((success) => success.data)
-      .finally(() => {
-        if (!opts.skipLoader) useAppSpinner.hide()
-        window.clearTimeout(wait)
-      })
-  },
+
+      if (err instanceof Error) {
+        throw new HttpError(err.message, 0)
+      }
+
+      throw new HttpError("An unknown error has occurred.", 0)
+    })
+    .finally(() => {
+      if (options.skipLoader !== true) {
+        useAppSpinner.hide()
+      }
+
+      window.clearTimeout(wait)
+    })
+}
+
+const BaseAPI: HttpClient = {
   get<T = any>(
     path: string,
-    opts: RequestOptions,
+    opts?: RequestOptions,
     params?: Record<string, unknown>
   ) {
-    return this.makeRequest<T>({ url: path, method: "GET", params }, opts)
+    return httpRequest<T>({ url: path, method: "GET", params }, opts || {})
+  },
+  delete<T = any>(path: string, opts?: RequestOptions) {
+    return httpRequest<T>({ url: path, method: "DELETE" }, opts || {})
   },
-  delete<T = any>(path: string, opts: RequestOptions) {
-    return this.makeRequest<T>({ url: path, method: "DELETE" }, opts)
+  patch<T = any>(
+    path: string,
+    data: Record<string, unknown> | FormData,
+    opts?: RequestOptions
+  ) {
+    return httpRequest<T>({ url: path, data, method: "PATCH" }, opts || {})
   },
   post<T = any>(
     path: string,
     data: Record<string, unknown> | FormData,
-    opts: RequestOptions
+    opts?: RequestOptions
   ) {
-    return this.makeRequest<T>({ url: path, data, method: "POST" }, opts)
+    return httpRequest<T>({ url: path, data, method: "POST" }, opts || {})
   },
   put<T = any>(
     path: string,
     data: Record<string, unknown> | FormData,
-    opts: RequestOptions
+    opts?: RequestOptions
   ) {
-    return this.makeRequest<T>(
+    return httpRequest<T>(
       {
         url: path,
         data,
         method: "PUT",
       },
-      opts
+      opts || {}
     )
   },
 }
 
 export default BaseAPI
+
+/**
+ * A convenience method for checking if a variable in a failed request has an http status code.
+ * This is most useful for checking for specific http status codes in error callbacks.
+ * @param err unknown
+ * @param codes number | number[]
+ * @returns boolean
+ */
+export const hasHttpErrStatus = (err: unknown, codes: number | number[]) => {
```

Comment:
Some amount of type narrowing should be used in catch blocks in favor of accepting `any`.  This makes it nice an easy to check an unknown variable for a status property.  `err?.status === 409` is okay, but will only pass type checks with an implicit/explicit any or an overly narrow assertion of `HttpError`.  Let's assert `HttpError | Error` and use this method for convenience when checking for status codes.

Reply:
I'm thinking about a dev 2 years from know looking at this function and curious why we aren't doing `err?.status === 409`. I'm not sure how to expose this info, but i'd love to see this live with the code in some way

Reply:
Yeah, this is a generally unpopular design decision in TypeScript it seems if you look at the number of 👎 tags on this [comment](https://github.com/microsoft/TypeScript/issues/33736).

One alternative, which may be better for clarity, is to use the type guard.  

```ts
UsersAPI.get(id)
  .then((result) => {
    user.value = result
  })
  .catch((err: HttpError | Error) => {
    if (isHttpError(err) && err.status === 409) {
      // do conflict handling
    }
  })
```

I think we had a few hairy ones on the portal side that pushed for more convenience and readability.
```ts
    if (isHttpError(err) && (err.status === http.statuses.BAD_REQUEST || http.statuses.INTERNAL_SERVER_ERROR)) {
      // do conflict and badhandling
    }
```
But there are many other ways to clean this up beyond a helper method like `hasHttpErrStatus`.

Reply:
I should note the `isHttpError` type guard was directly inspired by [Axios](https://github.com/axios/axios#typescript).